File: python_matlab_comparison/scripts/cpu/time_sweep_beamforming_cpu.m
%% some setup
if ~exist('pycom_path','var') % check for pycom path
    file_dir = fileparts(mfilename('fullpath')); %get the directory of this file
    pycom_path = fullfile(file_dir,'../../'); %assume the base is 2 levels up from this
end

if ~exist('DEBUG','var') % Check for debug flag
    DEBUG = false;
    fprintf("DEBUG Not defined. Setting to False");
end

if ~exist('output_directory','var') % check for provided output directory
    output_directory = './data';
end

addpath(fullfile(pycom_path,'python_matlab_comparison\beamforming')); % add OperationTimer
addpath(fullfile(pycom_path,'base'));

%% Some initialization
if DEBUG
    dim_list = floor(linspace(1,50,6));
    num_reps = 5;
else
    dim_list = floor(linspace(1,500,51));
    num_reps = 100;
end
out_file = fullfile(output_directory,'stats_mat_beamforming.mat');

%% Test the fancy_timeit_matrix_sweep capability
functs      = {@beamform_speed};
funct_names = {'beamforming'  };
num_args    = [2      ];

% now run 
stats_mat_double = OperationTimer.fancy_timeit_matrix_sweep(...
	functs,funct_names,num_args,dim_list,num_reps,'arg_gen_funct',@fdfd_arg_gen_funct);
stats_mat_single = OperationTimer.fancy_timeit_matrix_sweep(...
	functs,funct_names,num_args,dim_list,num_reps,'arg_gen_funct',@fdfd_arg_gen_funct_single);
	
%% now save out
m_single = stats_mat_single; % matlab cant do sparse single
m_double = stats_mat_double;
save(out_file,'m_single','m_double');
    
%% functions for fft arg generation                                     
function args = fdfd_arg_gen_funct(dim,num_args)
     %@brief generate default arguments
    args = {dim,@double};
end

function args = fdfd_arg_gen_funct_single(dim,num_args)
     %@brief generate default arguments
    args = {dim,@single};
end
    
    
    
    
	
	
	

File: FancyTimer/examples/scripts/cpu/time_sweep_beamforming_cpu.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov  5 16:14:54 2019

@author: aweiss
"""
import os
os.environ['NUMBA_NUM_THREADS'] = '12' # match the number used in MATLAB
from numba import config as nb_config
print(nb_config.NUMBA_NUM_THREADS)

from pycom.timing.OperationTimer import fancy_timeit_matrix_sweep
from pycom.timing.beamforming.beamform_speed import beamform_speed
import numpy as np
import scipy.linalg
import os

#this allows setting of a global debug
try:
    from __main__ import DEBUG
    if DEBUG is None:
        raise Exception("DEBUG is None")
except:
    DEBUG = False #debug mode for testing
    print("DEBUG Not defined. Setting to False")
if DEBUG:
    print("Debugging Mode Enabled")
    
# what directory to store the data in
try:
    from __main__ import output_directory
except:
    output_directory = './data'


#%% Some initialization
    
if DEBUG:
    dim_list = np.floor(np.linspace(1,50,6)).astype(np.uint32);
    num_reps = 10
else:
    dim_list = np.floor(np.linspace(1,500,51)).astype(np.uint32);
    num_reps = 100
    
out_file = os.path.join(output_directory,'stats_py_np_beamforming.mat')


stats_np_single = {}
stats_np_double = {}

#%% argument generation
def beamforming_arg_gen_funct(dim,num_args): #generate num_cells_x,num_cells_y as dim
    return [dim,np.cdouble]  
def beamforming_arg_gen_funct_single(dim,num_args): #generate num_cells_x,num_cells_y as dim
    return [dim,np.csingle]  

#%% now run
[solve_stats_double,_] = fancy_timeit_matrix_sweep(
        [beamform_speed],['beamforming'],[2],dim_list,num_reps
        ,arg_gen_funct=beamforming_arg_gen_funct);
[solve_stats_single,_] = fancy_timeit_matrix_sweep(
        [beamform_speed],['beamforming'],[2],dim_list,num_reps
        ,arg_gen_funct=beamforming_arg_gen_funct_single);
		
stats_np_double.update(solve_stats_double)
stats_np_single.update(solve_stats_single)
    
#%% now save out (single is not used)
import scipy.io as sio
sio.savemat(out_file,{'single':stats_np_single,'double':stats_np_double})
    